app/src/org/commcare/engine/cases/AndroidIndexedFixtureInstanceTreeElement.java

```diff
@@ -0,0 +1,80 @@
+package org.commcare.engine.cases;
+
+import org.commcare.cases.instance.IndexedFixtureInstanceTreeElement;
+import org.commcare.cases.model.StorageIndexedTreeElementModel;
+import org.commcare.core.interfaces.UserSandbox;
+import org.commcare.utils.SerializationUtil;
+import org.javarosa.core.model.IndexedFixtureIdentifier;
+import org.javarosa.core.model.instance.AbstractTreeElement;
+import org.javarosa.core.model.instance.InstanceBase;
+import org.javarosa.core.model.instance.TreeElement;
+import org.javarosa.core.services.storage.IStorageUtilityIndexed;
+
+/**
+ * The root element for an indexed fixture data instance
+ * with support for attributes lookup
+ */
+
+public class AndroidIndexedFixtureInstanceTreeElement extends IndexedFixtureInstanceTreeElement {
+
+    private TreeElement attributes;
+
+    public AndroidIndexedFixtureInstanceTreeElement(AbstractTreeElement instanceRoot,
+                                                    IStorageUtilityIndexed<StorageIndexedTreeElementModel> storage,
+                                                    IndexedFixtureIdentifier indexedFixtureIdentifier) {
+        super(instanceRoot, storage, indexedFixtureIdentifier);
+    }
+
+    public static IndexedFixtureInstanceTreeElement get(UserSandbox sandbox,
+                                                        String instanceName,
+                                                        InstanceBase instanceBase) {
+        IndexedFixtureIdentifier indexedFixtureIdentifier = sandbox.getIndexedFixtureIdentifier(instanceName);
+        if (indexedFixtureIdentifier == null) {
+            return null;
+        } else {
+            IStorageUtilityIndexed<StorageIndexedTreeElementModel> storage =
+                    sandbox.getIndexedFixtureStorage(instanceName);
+            return new AndroidIndexedFixtureInstanceTreeElement(instanceBase, storage, indexedFixtureIdentifier);
+        }
+    }
+
+    @Override
+    public int getAttributeCount() {
+        return getDeserializedAttributes().getAttributeCount();
+    }
+
+    @Override
+    public String getAttributeNamespace(int index) {
+        return getDeserializedAttributes().getAttributeNamespace(index);
+    }
+
+    @Override
+    public String getAttributeName(int index) {
+        return getDeserializedAttributes().getAttributeName(index);
+    }
+
+    @Override
+    public String getAttributeValue(int index) {
+        return getDeserializedAttributes().getAttributeValue(index);
+    }
+
+    @Override
+    public AbstractTreeElement getAttribute(String namespace, String name) {
+        TreeElement attr = getDeserializedAttributes().getAttribute(namespace, name);
+        attr.setParent(this);
+        return attr;
+    }
+
+    @Override
+    public String getAttributeValue(String namespace, String name) {
+        return getDeserializedAttributes().getAttributeValue(namespace, name);
+    }
+
+    private TreeElement getDeserializedAttributes() {
```

Comment:
This method needs to be synchronized in the same way as this one:
https://github.com/dimagi/commcare-core/blob/master/src/main/java/org/commcare/cases/instance/StorageInstanceTreeElement.java#L103

to ensure that access to the attributes class is properly managed between threads.

We may want to consider a similar approach to that class where we have one "lazy load method" (there it's called loadElements()) that's called upfront before any methods, rather than specifically presuming that the lazy load is around the attributes.

Reply:
Made the method synchronized here though I am not sure of any particular benefits of doing deserialization ahead of time, specially since it requires us to be mindful of not accessing the `attributes` before we call the load on them. For ex. even though `loadElements()` is designed to be called once at the start we still end up calling it at couple other places.

Reply:
Hi Shubham,

I think the lazy loading independent of attributes is just a stylistic thing, so fine for it to not be included.

-Clayton